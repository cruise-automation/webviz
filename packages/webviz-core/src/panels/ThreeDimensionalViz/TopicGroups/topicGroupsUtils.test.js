// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

/* eslint-disable jest/no-disabled-tests */
import {
  buildItemDisplayNameByTopicOrExtension,
  removeTopicPrefixes,
  buildAvailableNamespacesByTopic,
  getSelectionsFromTopicGroupConfig,
} from "./topicGroupsUtils";

const DEFAULT_TOPIC_CONFIG = {
  name: "root",
  children: [
    {
      name: "Ext A",
      extension: "ExtA.a",
      children: [
        {
          name: "Ext B",
          extension: "ExtB.b",
        },
        {
          name: "Ext C",
          extension: "ExtC.c",
        },
      ],
    },
    {
      name: "Some Topic in JSON Tree",
      topic: "/topic_in_json_tree",
    },
    {
      name: "TF",
      children: [],
      description: "Visualize relationships between /tf frames.",
    },
    {
      name: "Nested Group",
      children: [
        {
          name: "Topic A",
          topic: "/topic_a",
        },
        {
          name: "Topic B",
          topic: "/topic_b",
        },
        {
          name: "Deeply Nested Group",
          children: [{ topic: "/topic_c" }],
        },
        {
          topic: "/topic_d",
        },
      ],
    },
  ],
};

jest.mock("webviz-core/src/loadWebviz", () => ({
  getGlobalHooks: () => ({
    perPanelHooks: () => ({
      ThreeDimensionalViz: {
        getDefaultTopicTree: () => ({
          name: "root",
          children: [
            {
              name: "Ext A",
              extension: "ExtA.a",
              children: [
                {
                  name: "Ext B",
                  extension: "ExtB.b",
                },
                {
                  name: "Ext C",
                  extension: "ExtC.c",
                },
              ],
            },
            {
              name: "Some Topic in JSON Tree",
              topic: "/topic_in_json_tree",
            },
            {
              name: "TF",
              children: [],
              description: "Visualize relationships between /tf frames.",
            },
            {
              name: "Nested Group",
              children: [
                {
                  name: "Topic A",
                  topic: "/topic_a",
                },
                {
                  name: "Topic B",
                  topic: "/topic_b",
                },
                {
                  name: "Deeply Nested Group",
                  children: [{ topic: "/topic_c" }],
                },
                {
                  topic: "/topic_d",
                },
              ],
            },
          ],
        }),
      },
    }),
  }),
}));

describe("topicGroupUtils", () => {
  describe("removeTopicPrefixes", () => {
    it("removes webviz bag prefixes", () => {
      expect(
        removeTopicPrefixes([
          "/foo/bar",
          "/webviz_bag_2/foo",
          "/webviz_tables_2/some_table2_topic",
          "/webviz_labels/some_label_topic",
        ])
      ).toEqual(["/foo/bar", "/foo", "/some_table2_topic", "/some_label_topic"]);
    });
  });

  describe("buildItemDisplayNameByTopicOrExtension", () => {
    it("maps extension and topic to displayName", () => {
      expect(buildItemDisplayNameByTopicOrExtension(DEFAULT_TOPIC_CONFIG)).toEqual({
        "/metadata": "Map",
        "/tf": "TF",
        "/topic_a": "Nested Group / Topic A",
        "/topic_b": "Nested Group / Topic B",
        "/topic_c": "Nested Group / Deeply Nested Group",
        "/topic_d": "Nested Group",
        "/topic_in_json_tree": "Some Topic in JSON Tree",
        "ExtA.a": "Ext A",
        "ExtB.b": "Ext A / Ext B",
        "ExtC.c": "Ext A / Ext C",
      });
    });
  });
  describe("buildAvailableNamespacesByTopic", () => {
    it("builds a map of topics to namespaces including metadata, tf and topic namespaces", () => {
      expect(
        buildAvailableNamespacesByTopic({
          topicConfig: DEFAULT_TOPIC_CONFIG,
          allNamespaces: [
            { name: "some_ns1", topic: "/topic_a" },
            { name: "some_ns2", topic: "/topic_a" },
            { name: "some_ns1", topic: "/webviz_bag_2/topic_a" },
            { name: "some_ns1", topic: "/topic_c" },
          ],
          transformIds: ["some_tf_id1", "some_tf_id2"],
        })
      ).toEqual({
        "/metadata": ["ExtA.a", "ExtB.b", "ExtC.c"],
        "/tf": ["some_tf_id1", "some_tf_id2"],
        "/topic_a": ["some_ns1", "some_ns2"],
        "/topic_c": ["some_ns1"],
        "/webviz_bag_2/topic_a": ["some_ns1"],
      });
    });
  });

  // TODO(Audrey): finish the tests
  describe.skip("getTopicGroups", () => {
    it("generates topic group config for single bag", () => {});
    it("generates topic group config for two bags", () => {});
    it("generates topic group config for multiple data sources (bags + tables + labels)", () => {});
  });

  describe.skip("getNamespacesItemsBySource", () => {
    it("generates namespaces items with displayVisibilityBySource (hidden + available + badgeText)", () => {});
  });
  describe("getSelectionsFromTopicGroupConfig", () => {
    it("handles empty input", () => {
      expect(getSelectionsFromTopicGroupConfig([])).toEqual({
        selectedTopicNames: [],
        selectedNamespacesByTopic: {},
        selectedTopicSettingsByTopic: {},
      });
    });
    it("returns selectedTopicNames and selectedNamespacesByTopic", () => {
      expect(
        getSelectionsFromTopicGroupConfig([
          {
            displayName: "My Topic Group",
            visible: true,
            expanded: true,
            items: [
              // when missing visibilitiesBySource, select the non-prefixed topic as visible
              {
                topicName: "/topic_a",
              },
              {
                topicName: "/topic_b",
                visibilitiesBySource: { "/webviz_bag_2": false },
              },
              {
                topicName: "/topic_c",
                visibilitiesBySource: { "": true },
              },
              {
                topicName: "/topic_d",
                visibilitiesBySource: { "": false },
              },
              // omit topic if all data sources are not visible
              {
                topicName: "/topic_e",
                visibilitiesBySource: { "": false, "/webviz_bag_2": false },
              },
              {
                topicName: "/topic_f",
                visibilitiesBySource: { "": false, "/webviz_bag_2": true },
                selectedNamespacesBySource: {
                  "": ["some_ns11", "some_ns22"],
                  "/webviz_bag_2": ["some_ns33"],
                },
              },
              {
                topicName: "/topic_g",
                selectedNamespacesBySource: {
                  "": ["some_ns1", "some_ns2"],
                  "/webviz_bag_2": ["some_ns3"],
                },
                settingsBySource: {
                  "": {
                    overrideColor: "128, 0, 0, 1",
                    overrideCommand: "LinedConvexHull",
                  },
                  "/webviz_bag_2": {
                    overrideColor: "0, 128, 0, 1",
                    overrideCommand: "LinedConvexHull",
                  },
                },
              },
            ],
          },
          {
            displayName: "My Topic Group1",
            // all topics under this group won't be processed since the group is not visible
            visible: false,
            expanded: true,
            items: [
              {
                topicName: "/topic_h",
              },
            ],
          },
        ])
      ).toEqual({
        selectedNamespacesByTopic: {
          "/topic_f": ["some_ns11", "some_ns22"],
          "/webviz_bag_2/topic_f": ["some_ns33"],
          "/topic_g": ["some_ns1", "some_ns2"],
        },
        selectedTopicNames: ["/topic_a", "/topic_c", "/webviz_bag_2/topic_f", "/topic_g"],
        selectedTopicSettingsByTopic: {
          "/topic_g": {
            overrideColor: "128, 0, 0, 1",
            overrideCommand: "LinedConvexHull",
          },
          "/webviz_bag_2/topic_g": {
            overrideColor: "0, 128, 0, 1",
            overrideCommand: "LinedConvexHull",
          },
        },
      });
    });
  });
});
