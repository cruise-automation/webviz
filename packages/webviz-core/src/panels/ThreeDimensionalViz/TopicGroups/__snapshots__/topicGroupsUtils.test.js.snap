// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`topicGroupUtils getTopicGroups generates topic group config based on groupsConfig 1`] = `
Array [
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": true,
      "hasFeatureColumn": false,
      "id": "Some-Group1_0",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Group1",
    "items": Array [
      Object {
        "derivedFields": Object {
          "displayName": "/tf",
          "id": "Some-Group1_0_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/tf",
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": false,
      "hasFeatureColumn": false,
      "id": "Some-Group2_1",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Group2",
    "items": Array [
      Object {
        "derivedFields": Object {
          "displayName": "/some_topic1",
          "id": "Some-Group2_1_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/some_topic1",
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
]
`;

exports[`topicGroupUtils getTopicGroups generates topic group config for multiple data sources (bags + tables + labels) 1`] = `
Array [
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": true,
      "hasFeatureColumn": false,
      "id": "My-Topic-Group_0",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "My Topic Group",
    "expanded": true,
    "items": Array [
      Object {
        "derivedFields": Object {
          "datatype": "tf2_msgs/TFMessage",
          "displayName": "/tf",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": true,
            },
          ],
          "id": "My-Topic-Group_0_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "expanded": true,
        "selectedNamespacesByColumn": Array [
          Array [
            "some_tf_ns1",
          ],
          Array [
            "some_tf_ns1",
            "some_tf_ns2",
          ],
        ],
        "topicName": "/tf",
        "visibilityByColumn": Array [
          true,
          false,
        ],
      },
      Object {
        "derivedFields": Object {
          "displayName": "Map",
          "id": "My-Topic-Group_0_1",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "displayName": "Map",
        "topicName": "/metadata",
        "visibilityByColumn": Array [
          true,
          false,
        ],
      },
      Object {
        "derivedFields": Object {
          "displayName": "/some_topic_1",
          "id": "My-Topic-Group_0_2",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/some_topic_1",
        "visibilityByColumn": Array [
          true,
          true,
        ],
      },
      Object {
        "derivedFields": Object {
          "displayName": "/labels_json/some_label_topic_1",
          "id": "My-Topic-Group_0_3",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/labels_json_2",
          ],
        },
        "topicName": "/labels_json/some_label_topic_1",
      },
      Object {
        "derivedFields": Object {
          "availableNamespacesByColumn": Array [
            Array [
              "some_topic_2_ns1",
              "some_topic_2_ns2",
            ],
          ],
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "/some_topic_2",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "id": "My-Topic-Group_0_4",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
          "sortedNamespaceDisplayVisibilityByColumn": Array [
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": true,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "some_topic_2_ns1",
            },
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": false,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "some_topic_2_ns2",
            },
          ],
        },
        "expanded": true,
        "selectedNamespacesByColumn": Array [
          Array [
            "some_topic_2_ns1",
          ],
          Array [
            "some_topic_2_ns1",
            "some_topic_2_ns2",
          ],
        ],
        "settingsByColumn": Array [
          Object {
            "overrideColor": "128, 0, 0, 1",
            "overrideCommand": "LinedConvexHull",
          },
          Object {
            "overrideColor": "0, 128, 0, 1",
            "overrideCommand": "LinedConvexHull",
          },
        ],
        "topicName": "/some_topic_2",
        "visibilityByColumn": Array [
          false,
          true,
        ],
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": true,
      "hasFeatureColumn": false,
      "id": "Some-Topic-Group-1_1",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Topic Group 1",
    "expanded": true,
    "items": Array [
      Object {
        "derivedFields": Object {
          "datatype": "tf2_msgs/TFMessage",
          "displayName": "/tf",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "id": "Some-Topic-Group-1_1_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/tf",
      },
      Object {
        "derivedFields": Object {
          "availableNamespacesByColumn": Array [
            Array [
              "some_topic_2_ns1",
              "some_topic_2_ns2",
            ],
          ],
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "/some_topic_2",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "id": "Some-Topic-Group-1_1_1",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
          "sortedNamespaceDisplayVisibilityByColumn": Array [
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": true,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "some_topic_2_ns1",
            },
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": true,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "some_topic_2_ns2",
            },
          ],
        },
        "expanded": true,
        "topicName": "/some_topic_2",
      },
      Object {
        "derivedFields": Object {
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "/tables/some_topic",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "T1",
              "isParentVisible": false,
              "visible": true,
            },
          ],
          "id": "Some-Topic-Group-1_1_2",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_tables_2",
          ],
        },
        "settingsByColumn": Array [
          undefined,
          Object {
            "overrideColor": "0, 0, 255, 0.5",
            "overrideCommand": "LinedConvexHull",
          },
        ],
        "topicName": "/tables/some_topic",
        "visibilityByColumn": Array [
          true,
          true,
        ],
      },
      Object {
        "derivedFields": Object {
          "displayName": "/tables/some_topic_with_ns",
          "id": "Some-Topic-Group-1_1_3",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_tables_2",
          ],
        },
        "expanded": true,
        "topicName": "/tables/some_topic_with_ns",
        "visibilityByColumn": Array [
          true,
          true,
        ],
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
]
`;

exports[`topicGroupUtils getTopicGroups generates topic group config for two bags 1`] = `
Array [
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": true,
      "hasFeatureColumn": true,
      "id": "Some-Group_0",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Group",
    "items": Array [
      Object {
        "derivedFields": Object {
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "/some_topic1",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
            Object {
              "available": true,
              "badgeText": "B2",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "id": "Some-Group_0_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/some_topic1",
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
]
`;

exports[`topicGroupUtils getTopicGroups updates the groups based on available topics, namespaces, displayNames and errors 1`] = `
Array [
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": true,
      "hasFeatureColumn": true,
      "id": "Some-Group1_0",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Group1",
    "items": Array [
      Object {
        "derivedFields": Object {
          "availableNamespacesByColumn": Array [
            Array [
              "tf_ns1",
              "tf_ns2",
            ],
          ],
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "Transforms",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "id": "Some-Group1_0_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
          "sortedNamespaceDisplayVisibilityByColumn": Array [
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": true,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "tf_ns1",
            },
            Object {
              "displayVisibilityByColumn": Array [
                Object {
                  "available": true,
                  "badgeText": "B1",
                  "isParentVisible": false,
                  "visible": true,
                },
              ],
              "keyboardFocusIndex": -1,
              "namespace": "tf_ns2",
            },
          ],
        },
        "topicName": "/tf",
      },
      Object {
        "derivedFields": Object {
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "/some_topic2",
          "displayVisibilityByColumn": Array [
            undefined,
            Object {
              "available": true,
              "badgeText": "B2",
              "isParentVisible": false,
              "visible": false,
            },
          ],
          "errors": Array [
            "(/webviz_bag_2) missing frame id",
          ],
          "id": "Some-Group1_0_1",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/some_topic2",
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
  Object {
    "derivedFields": Object {
      "addTopicKeyboardFocusIndex": -1,
      "expanded": false,
      "hasFeatureColumn": true,
      "id": "Some-Group2_1",
      "isShownInList": true,
      "keyboardFocusIndex": -1,
      "prefixesByColumn": Array [
        Array [
          "",
          "/webviz_tables",
          "/labels_json",
        ],
        Array [
          "/webviz_bag_2",
          "/webviz_tables_2",
        ],
      ],
    },
    "displayName": "Some Group2",
    "items": Array [
      Object {
        "derivedFields": Object {
          "datatype": "visualization_msgs/MarkerArray",
          "displayName": "Some Topic 1",
          "displayVisibilityByColumn": Array [
            Object {
              "available": true,
              "badgeText": "B1",
              "isParentVisible": false,
              "visible": false,
            },
            undefined,
          ],
          "errors": Array [
            "missing transforms to my_root_frame",
          ],
          "id": "Some-Group2_1_0",
          "isShownInList": true,
          "keyboardFocusIndex": -1,
          "prefixByColumn": Array [
            "",
            "/webviz_bag_2",
          ],
        },
        "topicName": "/some_topic1",
      },
    ],
    "visibilityByColumn": Array [
      false,
      false,
    ],
  },
]
`;
